sql/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java

```diff
@@ -62,6 +65,15 @@ public void add(TItem item, ShardLocation shardLocation) {
         req.add(location, item);
     }
 
+    void addFailedItem(BytesRef sourceUri, String readFailure) {
+        List<String> itemsWithFailure = itemsWithFailureBySourceUri.computeIfAbsent(sourceUri, k -> new ArrayList<>());
+        itemsWithFailure.add(readFailure);
+    }
+
+    void addFailedUri(BytesRef sourceUri, String uriReadFailure) {
+        sourceUrisWithFailure.put(sourceUri, uriReadFailure);
```

Comment:
should we add a sanity assertion on the result of the `put` to make sure the entry shouldn't have existed?